shell/shared/db.js

```diff
@@ -86,6 +86,46 @@ Grains = new Mongo.Collection("grains");
 //   publicId:  An id used to publicly identify this grain. Used e.g. to route incoming e-mail and
 //       web publishing. This field is initialized when first requested by the app.
 
+RoleAssignments = new Mongo.Collection("roleAssignments");
+// A role assignment allows a user to access a grain with a specified set of permissions; it is an
+// edge in the permissions sharing graph. A user's permissions for a given grain are computed as
+// the union of the permissions received through role assignments.
+//
+// Any grain for which a user has received a role assignment will show up in that user's grain list.
+// Role assignments are also used to restrict access during restoration of persisted capabilities.
+//
+// Each contains:
+//   _id: random
+//   grainId: The `_id` of the grain whose permissions are being shared.
+//   sharer: The `_id` of the user who is sharing these permissions.
+//   recipient: The `_id` of the user who receives these permissions.
+//   permissions: An array of bools representing the received permissions.
+//   roleAssignment: A Grain.ViewSharingLink.RoleAssignment that is equivalent to `permissions`, but
+//                   may contain more semantic information. In JSON format.
+//   petname: Human-readable label chosen by and only visible to the sharer.
+//   title: Human-readable title as chosen by the recipient. Used in the same places that
+//          `grain.title` is used for the grain's owner.
+//   created: Date when this role assignment was created.
+//   expires: Optional expiration date. If undefined, the role assignment does not expire.
+//   parentKey: If present, the `_id` of the role assignment key from which this was derived.
```

Comment:
Hmm, how is `parentKey` supposed to be used? Say Alice shares to Carol who shares to Dave. Later, Alice revokes Carol's access, but now Bob separately shares the same role to Carol, restoring her access. Dave's access should be restored as well. So it seems the "parent" assignment has changed.

Or say Alice shares read-only access to Carol, and Bob shares write-only access to Carol, and Carol shares read/write access to Dave. Then there are two parents.

With that said, it may very well make sense to keep some sort of a cache of the authorization path used on a previous authorization check so that we don't have to do a full graph search each time.